src/cpp-common/include/cpp-common/HashMap.tpp

```diff
@@ -0,0 +1,145 @@
+            m_usedSpaces++;
+            m_usedSpacesFlag[index] = true;
+            return true;
+        }
+        if (m_usedSpacesFlag[index] == true) {
+            // If a key is already present in the map, do not overwrite its value
+            if (reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]).first ==
+                item.first) {
+                return false;
+            }
+        }
+
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            loopedOnce = true;
+            index = 0;
+        }
+    } while (index < maxSize);
+    return false;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::remove(Key key) {
+    uint16_t index = hash(key);
+    bool loopedOnce = false;
+    do {
+        if (m_usedSpacesFlag[index] == true) {
+            auto pair = reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]);
+            if (pair.first == key) {
+                pair.second.~MappedType();
+                m_usedSpacesFlag[index] = false;
+                return true;
+            }
+        }
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            index = 0;
+            loopedOnce = true;
+        }
+    } while (index < maxSize);
+    return {};
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+void HashMap<Key, MappedType, maxSize>::clear() {
+    for (int i = 0; i < maxSize; i++) {
+        // Could forego calling constructor since using placement new
+        reinterpret_cast<std::pair<Key, MappedType>*>(&m_storage[i])->second.~MappedType();
+        m_usedSpacesFlag[i] = false;
+    }
+    m_usedSpaces = 0;
+}
+
+template <typename Key, typename MappedType, uint16_t maxSize>
+bool HashMap<Key, MappedType, maxSize>::get(Key k, MappedType& item) const {
+    uint16_t index = hash(k);
+    bool loopedOnce = false;
+    do {
+        if (m_usedSpacesFlag[index] == true) {
+            auto pair = reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]);
+            if (pair.first == k) {
+                // emplace on reference passed
+                new (&item) MappedType(pair.second);
+                return true;
+            }
+        }
+        index++;
+        // Only loop across array once
+        if (!loopedOnce && index == maxSize) {
+            index = 0;
+            loopedOnce = true;
+        }
+    } while (index < maxSize);
+    return false;
+}
+template <typename Key, typename MappedType, uint16_t maxSize>
+std::optional<MappedType> HashMap<Key, MappedType, maxSize>::at(Key key) const {
+    uint16_t index = hash(key);
+    bool loopedOnce = false;
+    do {
+        if (m_usedSpacesFlag[index] == true) {
+            auto pair = reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]);
```

Comment:
You can use `auto& [first, second] = m_storage[index];` as of c++ 17, same for everywhere

Reply:
Tried it, had issues with so i'll keep the reinterpret cast.
For reference:
```
error: cannot decompose union type 'const std::aligned_storage<40, 8>::type' [clang-diagnostic-error]
```

Reply:
You still need to cast it as an std::pair sorry, the line would be 

`auto& [first, second] = reinterpret_cast<const std::pair<Key, MappedType>&>(m_storage[index]);`